src/julia_gc.cc

```diff
@@ -210,204 +126,471 @@ static inline void * align_ptr(void * p)
-        if (c == 0) {
-            treap_delete_node(treap);
-            return 1;
+        else {
+            node = node->right;
         }
-        else if (c < 0) {
-            treap = &L(*treap);
+        dealloc(del);
+        _nodes--;
+        if (alpha_hi * _nodes <= alpha_lo * _maxnodes) {
+            if (_root)
+                rebalance(_root, _nodes);
+            _maxnodes = _nodes;
         }
+    }
+    void remove_aux(Node *& node, T & item)
+    {
+        if (!node)
+            return;
+        int c = Cmp::cmp(item, node->item);
+        if (c < 0)
+            remove_aux(node->left, item);
+        else if (c > 0)
+            remove_aux(node->right, item);
         else {
-            treap = &R(*treap);
+            remove_node(node);
         }
     }
-    return 0;
-}
 
-static uint64_t xorshift_rng_state = 1;
+  public:
+    BalancedTree()
+    {
+        init();
+    }
+    static BalancedTree * make()
+    {
+        BalancedTree<T, Cmp> * result = alloc<BalancedTree<T, Cmp> >();
+        result->init();
+        return result;
+    }
+    static void destroy(BalancedTree<T, Cmp> * tree)
+    {
+        tree->deinit();
+        dealloc(tree);
+    }
+    ~BalancedTree()
+    {
+        remove_all();
+    }
+    void init()
+    {
+        _nodes = _maxnodes = 0;
+        _root = NULL;
+        if (!height_to_size_init) {
+            height_to_size_init = 1;
+            double w = 1.0;
+            for (int d = 0; d < MaxTreeDepth; d++) {
+                w *= (double)alpha_hi;
+                w /= (double)alpha_lo;
+                height_to_size[d] = (Int)w;
+            }
+        }
+    }
+    void deinit()
+    {
+        remove_all();
+    }
+    void remove_all()
+    {
+        delete_nodes(_root);
+    }
+    void insert(T item)
+    {
+        insert_aux(_root, item, 0);
+        if (_nodes > _maxnodes)
+            _maxnodes = _nodes;
+    }
+    T * find(T item)
+    {
+        return find_aux(_root, item);
+    }
+    void remove(T item)
+    {
+        remove_aux(_root, item);
+    }
+    Int depth_aux(Node * node)
+    {
+        if (node == NULL)
+            return 0;
+        Int m1 = depth_aux(node->left);
+        Int m2 = depth_aux(node->right);
+        return (m1 < m2 ? m2 : m1) + 1;
+    }
+    Int depth()
+    {
+        return depth_aux(_root);
+    }
+    Int count()
+    {
+        return _nodes;
+    }
+};
+
+struct CmpAddr {
+    static int cmp(void *a, void *b) {
+        return cmp_ptr(a, b);
+    }
+};
+
+
+} // namespace JuliaGC
+
+using namespace JuliaGC;
+
 
-static uint64_t xorshift_rng(void)
+
+#ifndef REQUIRE_PRECISE_MARKING
+
+#define MARK_CACHE_BITS 16
+#define MARK_CACHE_SIZE (1 << MARK_CACHE_BITS)
+
+#define MARK_HASH(x) (FibHash((x), MARK_CACHE_BITS))
+
+// The MarkCache exists to speed up the conservative tracing of
+// objects. While its performance benefit is minimal with the current
+// API functionality, it can significantly reduce overhead if a slower
```

Comment:
I don't want to sound bad, but if this provides very little benefit at present why add it, as it seems to be making things much more complicated.

Reply:
Yeah, the performance benefits vary. When I originally wrote it, I got like a 5%-10% speedup in some cases (for overall code, not just the GC part), but it seems to be compiler/architecture dependent as to whether it's actually a win. I'm fine with stripping that out in a separate PR if there aren't any performance benefits on common setups.

Reply:
To be clear: this PR does *not* add the MarkCache. It only moves it around. Unfortunately, GitHub's diff displaying logic fails to properly detect that a big chunk of code was inserted... But you can search for `#define MARK_CACHE_BITS` on this page to find the old place of the code (around line 80)